From 0d2b6c38f0cb682a0c209b2df17e7d08b18642e1 Mon Sep 17 00:00:00 2001
From: Rene Kita <mail@rkta.de>
Date: Tue, 12 Dec 2023 13:17:24 +0100
Subject: [PATCH 1/6] Enable building with "modern C" build options

Enable building with "modern C" build options

See the relevant Gentoo[0] and RedHat[1] bugs.

[0]: https://bugs.gentoo.org/show_bug.cgi?id=870412 an
[1]: https://bugzilla.redhat.com/show_bug.cgi?id=2142177

This enable to use
make OPTS="-Werror=implicit-function-declaration -Werror=implicit-int
-Werror=strict-prototypes"
---
 Str.c                     |    2 -
 buffer.c                  |    2 -
 display.c                 |   21 ++++++++++++++++----
 file.c                    |   12 +++++------
 fm.h                      |    2 -
 func.h                    |    2 -
 image.c                   |    8 +++----
 indep.c                   |    2 -
 istream.c                 |   16 +++++++--------
 istream.h                 |   24 +++++++++++------------
 libwc/wc_types.h          |   12 ++++++-----
 linein.c                  |    8 +++----
 local.c                   |    4 +--
 main.c                    |   10 ++++-----
 menu.c                    |    6 ++---
 menu.h                    |    2 -
 parsetagx.c               |   48 +++++++++++++++++++++++-----------------------
 proto.h                   |    4 +--
 rc.c                      |    2 -
 search.c                  |    2 -
 table.c                   |    2 -
 terms.c                   |   24 +++++++++++------------
 textlist.c                |    2 -
 w3mimg/fb/fb_gdkpixbuf.c  |    2 -
 w3mimg/fb/fb_img.h        |    2 -
 w3mimg/fb/fb_imlib2.c     |    2 -
 w3mimg/fb/fb_w3mimg.c     |    2 -
 w3mimg/w3mimg.c           |    2 -
 w3mimg/w3mimg.h           |    6 ++---
 w3mimg/win/win_w3mimg.cpp |    6 ++---
 w3mimg/x11/x11_w3mimg.c   |    2 -
 31 files changed, 128 insertions(+), 113 deletions(-)

--- a/buffer.c
+++ b/buffer.c
@@ -6,7 +6,7 @@
 #include <gpm.h>
 #endif
 #if defined(USE_GPM) || defined(USE_SYSMOUSE)
-extern int do_getch();
+extern int do_getch(void);
 #define getch()	do_getch()
 #endif				/* USE_GPM */
 #endif				/* USE_MOUSE */
--- a/display.c
+++ b/display.c
@@ -7,6 +7,19 @@
 /* *INDENT-OFF* */
 #ifdef USE_COLOR
 
+static void effect_anchor_start(void);
+static void effect_anchor_end(void);
+static void effect_image_start(void);
+static void effect_image_end(void);
+static void effect_form_start(void);
+static void effect_form_end(void);
+static void effect_active_start(void);
+static void effect_active_end(void);
+static void effect_visited_start(void);
+static void effect_visited_end(void);
+static void effect_mark_start(void);
+static void effect_mark_end(void);
+
 #define EFFECT_ANCHOR_START       effect_anchor_start()
 #define EFFECT_ANCHOR_END         effect_anchor_end()
 #define EFFECT_IMAGE_START        effect_image_start()
@@ -83,7 +96,7 @@ define_effect(EFFECT_MARK_START, EFFECT_
 
 /*****************/
 static void
-EFFECT_ACTIVE_START
+effect_active_start(void)
 {
     if (useColor) {
 	if (useActiveColor) {
@@ -104,7 +117,7 @@ EFFECT_ACTIVE_START
 }
 
 static void
-EFFECT_ACTIVE_END
+effect_active_end(void)
 {
     if (useColor) {
 	if (useActiveColor) {
@@ -118,7 +131,7 @@ EFFECT_ACTIVE_END
 }
 
 static void
-EFFECT_VISITED_START
+effect_visited_start(void)
 {
     if (useVisitedColor) {
 	if (useColor) {
@@ -130,7 +143,7 @@ EFFECT_VISITED_START
 }
 
 static void
-EFFECT_VISITED_END
+effect_visited_end(void)
 {
     if (useVisitedColor) {
 	if (useColor) {
--- a/file.c
+++ b/file.c
@@ -5588,7 +5588,7 @@ ex_efct(int ex)
 }
 
 static void
-HTMLlineproc2body(Buffer *buf, Str (*feed) (), int llimit)
+HTMLlineproc2body(Buffer *buf, Str (*feed) (void), int llimit)
 {
     static char *outc = NULL;
     static Lineprop *outp = NULL;
@@ -8193,7 +8193,7 @@ save2tmp(URLFile uf, char *tmpf)
     MySignalHandler(*volatile prevtrap) (SIGNAL_ARG) = NULL;
     static JMP_BUF env_bak;
     volatile int retval = 0;
-    char *volatile buf = NULL;
+    unsigned char *volatile buf = NULL;
 
     ff = fopen(tmpf, "wb");
     if (ff == NULL) {
@@ -8234,7 +8234,7 @@ save2tmp(URLFile uf, char *tmpf)
     {
 	int count;
 
-	buf = NewWithoutGC_N(char, SAVE_BUF_SIZE);
+	buf = NewWithoutGC_N(unsigned char, SAVE_BUF_SIZE);
 	while ((count = ISread_n(uf.stream, buf, SAVE_BUF_SIZE)) > 0) {
 	    if (fwrite(buf, 1, count, ff) != count) {
 		retval = -2;
@@ -8361,7 +8361,7 @@ _MoveFile(char *path1, char *path2)
     FILE *f2;
     int is_pipe;
     clen_t linelen = 0, trbyte = 0;
-    char *buf = NULL;
+    unsigned char *buf = NULL;
     int count;
 
     f1 = openIS(path1);
@@ -8380,7 +8380,7 @@ _MoveFile(char *path1, char *path2)
 	return -1;
     }
     current_content_length = 0;
-    buf = NewWithoutGC_N(char, SAVE_BUF_SIZE);
+    buf = NewWithoutGC_N(unsigned char, SAVE_BUF_SIZE);
     while ((count = ISread_n(f1, buf, SAVE_BUF_SIZE)) > 0) {
 	fwrite(buf, 1, count, f2);
 	linelen += count;
@@ -8747,7 +8747,7 @@ uncompress_stream(URLFile *uf, char **sr
 	}
 	if (pid2 == 0) {
 	    /* child2 */
-	    char *buf = NewWithoutGC_N(char, SAVE_BUF_SIZE);
+	    unsigned char *buf = NewWithoutGC_N(unsigned char, SAVE_BUF_SIZE);
 	    int count;
 	    FILE *f = NULL;
 
--- a/fm.h
+++ b/fm.h
@@ -1151,7 +1151,7 @@ global int relative_wheel_scroll init(FA
 global int fixed_wheel_scroll_count init(5);
 global int relative_wheel_scroll_ratio init(30);
 typedef struct _MouseActionMap {
-    void (*func) ();
+    void (*func) (void);
     char *data;
 } MouseActionMap;
 typedef struct _MouseAction {
--- a/func.h
+++ b/func.h
@@ -19,7 +19,7 @@
 
 typedef struct _FuncList {
     char *id;
-    void (*func) ();
+    void (*func) (void);
 } FuncList;
 
 #endif				/* not FUNC_H */
--- a/istream.c
+++ b/istream.c
@@ -20,7 +20,7 @@
 #define POP_CHAR(bs) ((bs)->iseos?'\0':(bs)->stream.buf[(bs)->stream.cur++])
 
 static int basic_close(int *handle);
-static int basic_read(int *handle, char *buf, int len);
+static int basic_read(int *handle, unsigned char *buf, int len);
 
 static int file_read(FILE *handle, char *buf, int len);
 
@@ -49,7 +49,7 @@ do_update(BaseStream base)
 }
 
 static int
-buffer_read(StreamBuffer sb, char *obuf, int count)
+buffer_read(StreamBuffer sb, unsigned char *obuf, int count)
 {
     int len = sb->next - sb->cur;
     if (len > 0) {
@@ -107,7 +107,7 @@ newInputStream(int des)
 }
 
 InputStream
-newFileStream(FILE * f, int (*closep) ())
+newFileStream(FILE * f, int (*closep)(FILE *))
 {
     InputStream stream;
     if (f == NULL)
@@ -178,7 +178,7 @@ newEncodedStream(InputStream is, char en
 int
 ISclose(InputStream stream)
 {
-    MySignalHandler(*prevtrap) ();
+    MySignalHandler(*prevtrap) (SIGNAL_ARG);
     if (stream == NULL)
         return -1;
     if (stream->base.close != NULL) {
@@ -291,7 +291,7 @@ ISread(InputStream stream, Str buf, int
 #endif
 
 int
-ISread_n(InputStream stream, char *dst, int count)
+ISread_n(InputStream stream, unsigned char *dst, int count)
 {
     int len, l;
     BaseStream base;
@@ -639,12 +639,12 @@ basic_close(int *handle)
 }
 
 static int
-basic_read(int *handle, char *buf, int len)
+basic_read(int *handle, unsigned char *buf, int len)
 {
 #ifdef __MINGW32_VERSION
-    return recv(*(int *)handle, buf, len, 0);
+    return recv(*handle, buf, len, 0);
 #else
-    return read(*(int *)handle, buf, len);
+    return read(*handle, buf, len);
 #endif
 }
 
--- a/istream.h
+++ b/istream.h
@@ -42,8 +42,8 @@ struct base_stream {
     void *handle;
     char type;
     char iseos;
-    int (*read) ();
-    int (*close) ();
+    int (*read) (int *, unsigned char *, int);
+    int (*close) (int *);
 };
 
 struct file_stream {
@@ -51,8 +51,8 @@ struct file_stream {
     FILE *handle;
     char type;
     char iseos;
-    int (*read) ();
-    int (*close) ();
+    int (*read) (FILE *, char *, int);
+    int (*close) (FILE *);
 };
 
 struct str_stream {
@@ -60,8 +60,8 @@ struct str_stream {
     Str handle;
     char type;
     char iseos;
-    int (*read) ();
-    int (*close) ();
+    int (*read) (Str, char *, int);
+    int (*close) (int);
 };
 
 #ifdef USE_SSL
@@ -70,8 +70,8 @@ struct ssl_stream {
     struct ssl_handle *handle;
     char type;
     char iseos;
-    int (*read) ();
-    int (*close) ();
+    int (*read) (struct ssl_handle *, char *, int);
+    int (*close) (struct ssl_handle *);
 };
 #endif				/* USE_SSL */
 
@@ -80,8 +80,8 @@ struct encoded_stream {
     struct ens_handle *handle;
     char type;
     char iseos;
-    int (*read) ();
-    int (*close) ();
+    int (*read) (struct ens_handle *, char *, int);
+    int (*close) (struct ens_handle *);
 };
 
 union input_stream {
@@ -105,7 +105,7 @@ typedef struct encoded_stream *EncodedSt
 typedef union input_stream *InputStream;
 
 extern InputStream newInputStream(int des);
-extern InputStream newFileStream(FILE * f, int (*closep) ());
+extern InputStream newFileStream(FILE * f, int (*closep) (FILE *));
 extern InputStream newStrStream(Str s);
 #ifdef USE_SSL
 extern InputStream newSSLStream(SSL * ssl, int sock);
@@ -121,7 +121,7 @@ void ISgets_to_growbuf(InputStream strea
 #ifdef unused
 extern int ISread(InputStream stream, Str buf, int count);
 #endif
-int ISread_n(InputStream stream, char *dst, int bufsize);
+int ISread_n(InputStream stream, unsigned char *dst, int bufsize);
 extern int ISfileno(InputStream stream);
 extern int ISeos(InputStream stream);
 #ifdef USE_SSL
--- a/libwc/wc_types.h
+++ b/libwc/wc_types.h
@@ -26,6 +26,8 @@ typedef wc_uint32 wc_ces;
 typedef wc_uint32 wc_locale;
 typedef wc_uchar  wc_bool;
 
+typedef struct wc_status wc_status;
+
 typedef struct {
     wc_ccs ccs;
     wc_uint32 code;
@@ -46,7 +48,7 @@ typedef struct {
     wc_ccs       ccs;
     size_t       n;
     wc_map      *map;
-    wc_wchar_t (*conv)();
+    wc_wchar_t (*conv)(wc_ccs, wc_uint16);
 } wc_table;
 
 typedef struct {
@@ -61,9 +63,9 @@ typedef struct {
     char     *desc;
     wc_gset  *gset;
     wc_uchar *gset_ext;
-    Str     (*conv_from)();
-    void    (*push_to)();
-    Str     (*char_conv)();
+    Str     (*conv_from)(Str, wc_ces);
+    void    (*push_to)(Str, wc_wchar_t, wc_status *);
+    Str     (*char_conv)(wc_uchar, wc_status *);
 } wc_ces_info;
 
 typedef struct {
@@ -94,7 +96,7 @@ typedef struct {
     wc_bool east_asian_width;	/* East Asian Ambiguous characters are wide */
 } wc_option;
 
-typedef struct {
+typedef struct wc_status {
     wc_ces_info *ces_info;
     wc_uint8     gr;
     wc_uint8     gl;
--- a/linein.c
+++ b/linein.c
@@ -8,7 +8,7 @@
 #include <gpm.h>
 #endif
 #if defined(USE_GPM) || defined(USE_SYSMOUSE)
-extern int do_getch();
+extern int do_getch(void);
 #define getch()	do_getch()
 #endif				/* USE_GPM */
 #endif				/* USE_MOUSE */
@@ -43,14 +43,14 @@ static int getcntrl(void);
 #endif
 
 static int terminated(unsigned char c);
-#define iself ((void(*)())insertself)
+#define iself ((void(*)(void))insertself)
 
 static void next_compl(int next);
 static void next_dcompl(int next);
 static Str doComplete(Str ifn, int *status, int next);
 
 /* *INDENT-OFF* */
-void (*InputKeymap[32]) () = {
+void (*InputKeymap[32]) (void) = {
 /*  C-@     C-a     C-b     C-c     C-d     C-e     C-f     C-g     */
     _compl, _mvB,   _mvL,   _inbrk, delC,   _mvE,   _mvR,   _inbrk,
 /*  C-h     C-i     C-j     C-k     C-l     C-m     C-n     C-o     */
@@ -208,7 +208,7 @@ inputLineHistSearch(char *prompt, char *
 	else if (!i_quote && c < 0x20) {	/* Control code */
 	    if (incrfunc == NULL
 		|| (c = incrfunc((int)c, strBuf, strProp)) < 0x20)
-		(*InputKeymap[(int)c]) (c);
+		(*InputKeymap[(int)c]) ();
 	    if (incrfunc && c != (unsigned char)-1 && c != CTRL_J)
 		incrfunc(-1, strBuf, strProp);
 	    if (cm_clear)
--- a/local.c
+++ b/local.c
@@ -28,7 +28,7 @@ static Str Local_cookie = NULL;
 static char *Local_cookie_file = NULL;
 
 static void
-writeLocalCookie()
+writeLocalCookie(void)
 {
     FILE *f;
 
@@ -47,7 +47,7 @@ writeLocalCookie()
 
 /* setup cookie for local CGI */
 Str
-localCookie()
+localCookie(void)
 {
     if (Local_cookie)
 	return Local_cookie;
--- a/main.c
+++ b/main.c
@@ -32,7 +32,7 @@
 #include <gpm.h>
 #endif				/* USE_GPM */
 #if defined(USE_GPM) || defined(USE_SYSMOUSE)
-extern int do_getch();
+extern int do_getch(void);
 #define getch()	do_getch()
 #endif				/* defined(USE_GPM) || defined(USE_SYSMOUSE) */
 #endif
@@ -1529,7 +1529,7 @@ static Str currentURL(void);
 
 #ifdef USE_BUFINFO
 void
-saveBufferInfo()
+saveBufferInfo(void)
 {
     FILE *fp;
 
@@ -1787,7 +1787,7 @@ clear_mark(Line *l)
 static int
 srchcore(char *volatile str, int (*func) (Buffer *, char *))
 {
-    MySignalHandler(*prevtrap) ();
+    MySignalHandler(*prevtrap) (SIGNAL_ARG);
     volatile int i, result = SR_NOTFOUND;
 
     if (str != NULL && str != SearchString)
@@ -2192,7 +2192,7 @@ DEFUN(pipesh, PIPE_SHELL, "Execute shell
 DEFUN(readsh, READ_SHELL, "Execute shell command and display output")
 {
     Buffer *buf;
-    MySignalHandler(*prevtrap) ();
+    MySignalHandler(*prevtrap) (SIGNAL_ARG);
     char *cmd;
 
     CurrentKeyData = NULL;	/* not allowed in w3m-control: */
@@ -5989,7 +5989,7 @@ getCodePage(void)
 #endif
 
 void
-deleteFiles()
+deleteFiles(void)
 {
     Buffer *buf;
     char *f;
--- a/menu.c
+++ b/menu.c
@@ -24,7 +24,7 @@ extern int sysm_process_mouse(int, int,
 #if defined(USE_GPM) || defined(USE_SYSMOUSE)
 #define X_MOUSE_SELECTED (char)0xff
 static int X_Mouse_Selection;
-extern int do_getch();
+extern int do_getch(void);
 #define getch()	do_getch()
 #endif				/* defined(USE_GPM) || defined(USE_SYSMOUSE) */
 #endif				/* USE_MOUSE */
@@ -666,7 +666,7 @@ guess_menu_xy(Menu *parent, int width, i
 }
 
 void
-new_option_menu(Menu *menu, char **label, int *variable, void (*func) ())
+new_option_menu(Menu *menu, char **label, int *variable, void (*func) (void))
 {
     int i, nitem;
     char **p;
@@ -1653,7 +1653,7 @@ smDelTab(char c)
 
 void
 optionMenu(int x, int y, char **label, int *variable, int initial,
-	   void (*func) ())
+	   void (*func) (void))
 {
     Menu menu;
 
--- a/menu.h
+++ b/menu.h
@@ -21,7 +21,7 @@ typedef struct _MenuItem {
     char *label;
     int *variable;
     int value;
-    void (*func) ();
+    void (*func) (void);
     struct _Menu *popup;
     char *keys;
     char *data;
--- a/parsetagx.c
+++ b/parsetagx.c
@@ -10,14 +10,14 @@
 
 /* parse HTML tag */
 
-static int noConv(char *, char **);
-static int toNumber(char *, int *);
-static int toLength(char *, int *);
-static int toAlign(char *, int *);
-static int toVAlign(char *, int *);
+static int noConv(char *, void *);
+static int toNumber(char *, void *);
+static int toLength(char *, void *);
+static int toAlign(char *, void *);
+static int toVAlign(char *, void *);
 
 /* *INDENT-OFF* */
-static int (*toValFunc[]) () = {
+static int (*toValFunc[]) (char *, void *) = {
     noConv,		/* VTYPE_NONE    */
     noConv,		/* VTYPE_STR     */
     toNumber,		/* VTYPE_NUMBER  */
@@ -33,14 +33,14 @@ static int (*toValFunc[]) () = {
 /* *INDENT-ON* */
 
 static int
-noConv(char *oval, char **str)
+noConv(char *oval, void *str)
 {
-    *str = oval;
+    *(char **)str = oval;
     return 1;
 }
 
 static int
-toNumber(char *oval, int *num)
+toNumber(char *oval, void *num)
 {
     char *ep;
     int x;
@@ -48,7 +48,7 @@ toNumber(char *oval, int *num)
     x = strtol(oval, &ep, 10);
 
     if (ep > oval) {
-	*num = x;
+	*(int *)num = x;
 	return 1;
     }
     else
@@ -56,7 +56,7 @@ toNumber(char *oval, int *num)
 }
 
 static int
-toLength(char *oval, int *len)
+toLength(char *oval, void *len)
 {
     int w;
     if (!IS_DIGIT(oval[0]))
@@ -67,41 +67,41 @@ toLength(char *oval, int *len)
     if (w == 0)
 	w = 1;
     if (oval[strlen(oval) - 1] == '%')
-	*len = -w;
+	*(int *)len = -w;
     else
-	*len = w;
+	*(int *)len = w;
     return 1;
 }
 
 static int
-toAlign(char *oval, int *align)
+toAlign(char *oval, void *align)
 {
     if (strcasecmp(oval, "left") == 0)
-	*align = ALIGN_LEFT;
+	*(int *)align = ALIGN_LEFT;
     else if (strcasecmp(oval, "right") == 0)
-	*align = ALIGN_RIGHT;
+	*(int *)align = ALIGN_RIGHT;
     else if (strcasecmp(oval, "center") == 0)
-	*align = ALIGN_CENTER;
+	*(int *)align = ALIGN_CENTER;
     else if (strcasecmp(oval, "top") == 0)
-	*align = ALIGN_TOP;
+	*(int *)align = ALIGN_TOP;
     else if (strcasecmp(oval, "bottom") == 0)
-	*align = ALIGN_BOTTOM;
+	*(int *)align = ALIGN_BOTTOM;
     else if (strcasecmp(oval, "middle") == 0)
-	*align = ALIGN_MIDDLE;
+	*(int *)align = ALIGN_MIDDLE;
     else
 	return 0;
     return 1;
 }
 
 static int
-toVAlign(char *oval, int *valign)
+toVAlign(char *oval, void *valign)
 {
     if (strcasecmp(oval, "top") == 0 || strcasecmp(oval, "baseline") == 0)
-	*valign = VALIGN_TOP;
+	*(int *)valign = VALIGN_TOP;
     else if (strcasecmp(oval, "bottom") == 0)
-	*valign = VALIGN_BOTTOM;
+	*(int *)valign = VALIGN_BOTTOM;
     else if (strcasecmp(oval, "middle") == 0)
-	*valign = VALIGN_MIDDLE;
+	*(int *)valign = VALIGN_MIDDLE;
     else
 	return 0;
     return 1;
--- a/proto.h
+++ b/proto.h
@@ -777,7 +777,7 @@ extern int action_menu(Menu *menu);
 extern void popup_menu(Menu *parent, Menu *menu);
 extern void guess_menu_xy(Menu *menu, int width, int *x, int *y);
 extern void new_option_menu(Menu *menu, char **label, int *variable,
-			    void (*func) ());
+			    void (*func) (void));
 
 extern int setMenuItem(MenuItem *item, char *type, char *line);
 extern int addMenuList(MenuList **list, char *id);
@@ -789,7 +789,7 @@ extern void mainMn(void);
 extern void selMn(void);
 extern void tabMn(void);
 extern void optionMenu(int x, int y, char **label, int *variable, int initial,
-		       void (*func) ());
+		       void (*func) (void));
 extern void initMenu(void);
 #else				/* not USE_MENU */
 #define mainMn nulcmd
--- a/rc.c
+++ b/rc.c
@@ -811,7 +811,7 @@ compare_table(struct rc_search_table *a,
 }
 
 static void
-create_option_search_table()
+create_option_search_table(void)
 {
     int i, j, k;
     int diff1, diff2;
--- a/terms.c
+++ b/terms.c
@@ -2097,7 +2097,7 @@ clrtoeolx(void)
 #endif				/* not USE_BG_COLOR */
 
 static void
-clrtobot_eol(void (*clrtoeol) ())
+clrtobot_eol(void (*clrtoeol) (void))
 {
     int l, c;
 
@@ -2359,7 +2359,7 @@ wgetch(void *p)
 }
 
 int
-do_getch()
+do_getch(void)
 {
     if (is_xterm || !gpm_handler)
 	return getch();
@@ -2370,7 +2370,7 @@ do_getch()
 
 #ifdef USE_SYSMOUSE
 int
-sysm_getch()
+sysm_getch(void)
 {
     fd_set rfd;
     int key, x, y;
@@ -2390,7 +2390,7 @@ sysm_getch()
 }
 
 int
-do_getch()
+do_getch(void)
 {
     if (is_xterm || !sysm_handler)
 	return getch();
@@ -2492,7 +2492,7 @@ sleep_till_anykey(int sec, int purge)
 /* Linux console with GPM support */
 
 void
-mouse_init()
+mouse_init(void)
 {
     Gpm_Connect conn;
     extern int gpm_process_mouse(Gpm_Event *, void *);
@@ -2529,7 +2529,7 @@ mouse_init()
 }
 
 void
-mouse_end()
+mouse_end(void)
 {
     if (mouseActive == 0)
 	return;
@@ -2544,7 +2544,7 @@ mouse_end()
 #elif	defined(USE_SYSMOUSE)
 /* *BSD console with sysmouse support */
 void
-mouse_init()
+mouse_init(void)
 {
     mouse_info_t mi;
     extern int sysm_process_mouse();
@@ -2585,7 +2585,7 @@ mouse_init()
 }
 
 void
-mouse_end()
+mouse_end(void)
 {
     if (mouseActive == 0)
 	return;
@@ -2606,7 +2606,7 @@ mouse_end()
 /* not GPM nor SYSMOUSE, but use mouse with xterm */
 
 void
-mouse_init()
+mouse_init(void)
 {
     if (mouseActive)
 	return;
@@ -2622,7 +2622,7 @@ mouse_init()
 }
 
 void
-mouse_end()
+mouse_end(void)
 {
     if (mouseActive == 0)
 	return;
@@ -2641,14 +2641,14 @@ mouse_end()
 
 
 void
-mouse_active()
+mouse_active(void)
 {
     if (!mouseActive)
 	mouse_init();
 }
 
 void
-mouse_inactive()
+mouse_inactive(void)
 {
     if (mouseActive && is_xterm)
 	mouse_end();
--- a/w3mimg/w3mimg.h
+++ b/w3mimg/w3mimg.h
@@ -44,16 +44,16 @@ typedef struct _w3mimg_op {
 } w3mimg_op;
 
 #ifdef USE_W3MIMG_X11
-extern w3mimg_op *w3mimg_x11open();
+extern w3mimg_op *w3mimg_x11open(void);
 #endif
 #ifdef USE_W3MIMG_FB
-extern w3mimg_op *w3mimg_fbopen();
+extern w3mimg_op *w3mimg_fbopen(void);
 #endif
 #ifdef USE_W3MIMG_WIN
 extern w3mimg_op *w3mimg_winopen();
 #endif
 
-extern w3mimg_op *w3mimg_open();
+extern w3mimg_op *w3mimg_open(void);
 
 #ifdef __cplusplus
 }
--- a/Str.c
+++ b/Str.c
@@ -34,7 +34,7 @@
 #endif				/* not STR_DEBUG */
 
 Str
-Strnew()
+Strnew(void)
 {
     Str x = GC_MALLOC(sizeof(struct _Str));
     if (x == NULL)
--- a/image.c
+++ b/image.c
@@ -36,7 +36,7 @@ static void closeImgdisplay(void);
 static int getCharSize(void);
 
 void
-initImage()
+initImage(void)
 {
     if (activeImage)
 	return;
@@ -95,7 +95,7 @@ getCharSize(void)
 }
 
 void
-termImage()
+termImage(void)
 {
     if (!activeImage)
 	return;
@@ -108,7 +108,7 @@ termImage()
 }
 
 static int
-openImgdisplay()
+openImgdisplay(void)
 {
     char *cmd;
 
@@ -298,7 +298,7 @@ drawImage(void)
 }
 
 void
-clearImage()
+clearImage(void)
 {
     static char buf[64];
     int j;
--- a/indep.c
+++ b/indep.c
@@ -123,7 +123,7 @@ strCmp(const void *s1, const void *s2)
 }
 
 char *
-currentdir()
+currentdir(void)
 {
     char *path;
 #ifdef HAVE_GETCWD
--- a/search.c
+++ b/search.c
@@ -19,7 +19,7 @@ static int migemo_running;
 static int migemo_pid = 0;
 
 void
-init_migemo()
+init_migemo(void)
 {
     migemo_active = migemo_running = use_migemo;
     if (migemor != NULL)
--- a/table.c
+++ b/table.c
@@ -269,7 +269,7 @@ table_border_width(struct table *t)
 }
 
 struct table *
-newTable()
+newTable(void)
 {
     struct table *t;
     int i, j;
--- a/textlist.c
+++ b/textlist.c
@@ -18,7 +18,7 @@ newListItem(void *s, ListItem *n, ListIt
 }
 
 GeneralList *
-newGeneralList()
+newGeneralList(void)
 {
     GeneralList *tl = New(GeneralList);
     tl->first = tl->last = NULL;
--- a/w3mimg/fb/fb_gdkpixbuf.c
+++ b/w3mimg/fb/fb_gdkpixbuf.c
@@ -81,7 +81,7 @@ get_animation_size(GdkPixbufAnimation *
 #endif
 
 void
-fb_image_init()
+fb_image_init(void)
 {
 #if defined(USE_GTK2)
     g_type_init();
--- a/w3mimg/fb/fb_img.h
+++ b/w3mimg/fb/fb_img.h
@@ -3,7 +3,7 @@
 #define fb_img_header
 #include "fb.h"
 
-void fb_image_init();
+void fb_image_init(void);
 FB_IMAGE **fb_image_load(char *filename, int w, int h, int n);
 void fb_image_set_bg(int r, int g, int b);
 int fb_image_clear(int x, int y, int w, int h);
--- a/w3mimg/fb/fb_imlib2.c
+++ b/w3mimg/fb/fb_imlib2.c
@@ -11,7 +11,7 @@ static void draw(FB_IMAGE * img, Imlib_I
 static Imlib_Image resize_image(Imlib_Image image, int width, int height);
 
 void
-fb_image_init()
+fb_image_init(void)
 {
     return;
 }
--- a/w3mimg/fb/fb_w3mimg.c
+++ b/w3mimg/fb/fb_w3mimg.c
@@ -169,7 +169,7 @@ check_tty_console(char *tty)
 #endif
 
 w3mimg_op *
-w3mimg_fbopen()
+w3mimg_fbopen(void)
 {
     w3mimg_op *wop = NULL;
     wop = (w3mimg_op *) malloc(sizeof(w3mimg_op));
--- a/w3mimg/w3mimg.c
+++ b/w3mimg/w3mimg.c
@@ -7,7 +7,7 @@
 #include "w3mimg/w3mimg.h"
 
 w3mimg_op *
-w3mimg_open()
+w3mimg_open(void)
 {
     w3mimg_op *w_op = NULL;
 #ifdef W3MIMGDISPLAY_SETUID
--- a/w3mimg/win/win_w3mimg.cpp
+++ b/w3mimg/win/win_w3mimg.cpp
@@ -97,11 +97,11 @@ static char *wstr2mb_alloc(const WCHAR *
     struct win_info *xi = static_cast<struct win_info *>(wop->priv); \
     assert(xi)
 
-win_image::win_image() THROW_NONE
+win_image::win_image(void) THROW_NONE
     : gpbitmap(NULL), nframe(0)
 {}
 
-win_image::~win_image() THROW_NONE
+win_image::~win_image(void) THROW_NONE
 {
     if (this->cache) {
 	for (size_t i = 0; i != this->nframe; ++i) {
@@ -680,7 +680,7 @@ last:
 }
 
 extern "C" w3mimg_op *
-w3mimg_winopen()
+w3mimg_winopen(void)
 {
     w3mimg_op *retval = NULL;
     Gdiplus::Status status = Gdiplus::Ok;
--- a/w3mimg/x11/x11_w3mimg.c
+++ b/w3mimg/x11/x11_w3mimg.c
@@ -798,7 +798,7 @@ x11_get_image_size(w3mimg_op * self, W3M
 /* *INDENT-ON* */
 
 w3mimg_op *
-w3mimg_x11open()
+w3mimg_x11open(void)
 {
     w3mimg_op *wop = NULL;
     struct x11_info *xi = NULL;
